src/storage/store/node_table.cpp

```diff
@@ -147,16 +147,41 @@ offset_t NodeTable::validateUniquenessConstraint(const Transaction* transaction,
     return INVALID_OFFSET;
 }
 
+static std::string getPkStr(const ValueVector& pkVector, sel_t pkPos) {
+    std::string pkStr;
+    TypeUtils::visit(
+        pkVector.dataType.getPhysicalType(),
+        [&](ku_string_t) { pkStr = pkVector.getValue<ku_string_t>(pkPos).getAsString(); },
+        [&pkStr, &pkVector, &pkPos]<typename T>(
+            T) { pkStr = TypeUtils::toString(pkVector.getValue<T>(pkPos)); });
+    return pkStr;
+}
+
+void NodeTable::validatePkNotExists(const Transaction* transaction, ValueVector* pkVector) {
+    common::offset_t dummyOffset;
+    auto& selVector = pkVector->state->getSelVector();
+    KU_ASSERT(selVector.getSelSize() == 1);
+    if (pkVector->isNull(selVector[0])) {
+        throw RuntimeException(ExceptionMessage::nullPKException());
+    }
+    if (pkIndex->lookup(transaction, pkVector, selVector[0], dummyOffset,
+            [&](offset_t offset) { return isVisible(transaction, offset); })) {
+        throw RuntimeException(
+            ExceptionMessage::duplicatePKException(getPkStr(*pkVector, selVector[0])));
```

Comment:
Alternative option to get the string here is going through `ValueVector::getAsValue()`, which I think it's a better way.